pkg/serviceaccount/jwt.go

```diff
@@ -225,22 +225,59 @@ func (j *jwtTokenGenerator) GenerateToken(claims *jwt.Claims, privateClaims inte
 // JWTTokenAuthenticator authenticates tokens as JWT tokens produced by JWTTokenGenerator
 // Token signatures are verified using each of the given public keys until one works (allowing key rotation)
 // If lookup is true, the service account and secret referenced as claims inside the token are retrieved and verified with the provided ServiceAccountTokenGetter
-func JWTTokenAuthenticator(issuers []string, keys []interface{}, implicitAuds authenticator.Audiences, validator Validator) authenticator.Token {
+func JWTTokenAuthenticator(issuers []string, keyGetter PublicKeyGetter, implicitAuds authenticator.Audiences, validator Validator) authenticator.Token {
 	issuersMap := make(map[string]bool)
 	for _, issuer := range issuers {
 		issuersMap[issuer] = true
 	}
 	return &jwtTokenAuthenticator{
 		issuers:      issuersMap,
-		keys:         keys,
+		keys:         keyGetter,
 		implicitAuds: implicitAuds,
 		validator:    validator,
 	}
 }
 
+// PublicKeyGetter returns public keys for a given key id.
+type PublicKeyGetter interface {
+	// GetPublicKeys returns public keys to use for verifying a token with the given key id.
+	// The token keyID may be empty if the token did not have a key id header.
+	GetPublicKeys(keyID string) []interface{}
+}
+
+type staticPublicKeyGetter struct {
+	publicKeys     []interface{}
+	publicKeysByID map[string][]interface{}
+}
+
+// StaticPublicKeyGetter constructs an implementation of PublicKeyGetter
+// which returns all public keys when key id is unspecified, and returns
+// the public keys matching the keyIDFromPublicKey-derived key id when
+// a key id is specified.
+func StaticPublicKeyGetter(publicKeys []interface{}) *staticPublicKeyGetter {
+	publicKeysByID := map[string][]interface{}{}
+	for _, key := range publicKeys {
+		keyID, err := keyIDFromPublicKey(key)
```

Comment:
I assume we have never changed how `keyIDFromPublicKey` works?  Because we did not check key ID on verification before, we could have changed this function over time without breaking validation since all keys were always tried.  Now we are requiring any non-empty encoded key ID to match on verification, which means that any legacy SA tokens with a non-empty key ID need to have had their key ID generated from this function.  It seems like this function should be marked as part of our public API, along with tests to ensure that it cannot change.

Reply:
added in https://github.com/kubernetes/kubernetes/commit/b4e99584ce4327324736e98a46ee57a626a3b96f#diff-4a2c2c85264a6a52b3e2e7524c04eaf49e638eea2b819b799e411fb2edb58efbR96-R109 and unchanged since then

will hoist this to a spot where it's clearer this is API surface and add stability tests to ensure we don't change it in the future

Reply:
added stability test with comments